pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -210,19 +231,134 @@ pub(super) fn spawn_storage_sync_thread<
     Ok(handle)
 }
 
-fn storage_sync_loop<P: std::fmt::Debug, S: 'static + RemoteStorage<StoragePath = P>>(
+fn storage_sync_loop<
+    P: std::fmt::Debug + Send + Sync + 'static,
+    S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
+>(
     config: &'static PageServerConf,
     remote_storage: S,
-    concurrent_sync_limit: &Semaphore,
+    max_concurrent_sync: NonZeroUsize,
+    max_sync_errors: NonZeroU32,
 ) -> anyhow::Result<()> {
     let runtime = tokio::runtime::Builder::new_current_thread()
         .enable_all()
         .build()?;
-    let mut remote_timelines = runtime
+    let remote_timelines = runtime
         .block_on(fetch_existing_uploads(&remote_storage))
         .context("Failed to determine previously uploaded timelines")?;
 
-    let urgent_downloads = latest_timelines(&remote_timelines)
+    schedule_first_tasks(config, &remote_timelines);
+
+    let remote_timelines = Arc::new(Mutex::new(remote_timelines));
+    let remote_storage = Arc::new(remote_storage);
+
+    while !crate::tenant_mgr::shutdown_requested() {
+        let next_tasks = get_next_tasks(max_concurrent_sync);
+        let remaining_queue_length = SYNC_QUEUE.len();
+
+        if next_tasks.is_empty() {
+            trace!("No storage sync tasks found");
+            thread::sleep(Duration::from_secs(1));
+            continue;
```

Comment:
I think `std::sync::mpsc` might make more sense then `SegQueue`, as it has blocking behavior

Reply:
How could one implement batching with such model?
It appears beneficial to have one, since allows us to reorder or merge some tasks, but there's no way to know without blocking/awaiting (I think it's better to be the tokio version of the channel in the async code, but not important) if there are any more tasks left in the channel's buffer?

Reply:
`recv` for the first task, then `try_recv` for the following tasks.

Reply:
Ok, had come up with the changes to reflect that.
There's a possibility to drop any of the ends to stop the whole machinery, but I have not found a good way to plug that into our loop concept. In theory, I could place the whole mpsc under the lock, but that beats the purpose of the whole idea IMO.
It might be better, if we move from static functions to some wrapper for the uploaded, as it was in the very first version, but that's a separate step, I think.